src/kmod/hash-table-generic.c

```diff
@@ -7,62 +7,131 @@
+	struct hlist_node *_node;
+	hlist_for_each_entry_safe (node, _node, tmp, head, link) {
+#else
+	hlist_for_each_entry_safe (node, tmp, head, link) {
+#endif
+		if (hash == node->hash &&
+		    hashTblp->cmp_key(key,
+				      get_key_ptr(hashTblp,
+						  get_datap(hashTblp, node)),
+				      hashTblp->cmp_len)) {
+			return node;
+		}
 	}
-	str[key_len * 3 - 1] = '\0';
-	return str;
+	return NULL;
 }
 
-inline void *get_key_ptr(struct HashTbl *hashTblp, void *datap)
+static bool cmp_key_u8(const void *a, const void *b, int key_len)
 {
-	return (void *)datap + hashTblp->key_offset;
+	return memcmp(a, b, key_len) == 0;
+}
+static bool cmp_key_u32(const void *a, const void *b, int indices)
+{
+	int		i;
+	const uint32_t *a32 = (const uint32_t *)a;
+	const uint32_t *b32 = (const uint32_t *)b;
+
+	for (i = 0; i < indices; i++, a32++, b32++) {
+		if (*a32 != *b32) {
+			return false;
+		}
+	}
+	return true;
 }
 
+static int debug = 0;
+
+uint64_t g_hashtbl_generic_lock = 0;
+LIST_HEAD(g_hashtbl_generic);
+
+#define HASHTBL_PRINT(fmt, ...)                            \
+	if (debug) {                                       \
+		PR_DEBUG("hash-tbl: " fmt, ##__VA_ARGS__); \
+	}
+
+static void __hashtbl_for_each_generic(struct HashTbl *,
+				       hashtbl_for_each_generic_cb, void *);
+
 struct HashTbl *hashtbl_init_generic(uint64_t numberOfBuckets,
 				     uint64_t datasize, uint64_t sizehint,
 				     const char *hashtble_name, int key_len,
 				     int key_offset, int node_offset)
 {
-	struct HashTbl *hashTblp = NULL;
-	int tableSize = ((numberOfBuckets * sizeof(struct hlist_head)) +
-			 sizeof(struct HashTbl));
-	unsigned char *tbl_storage_p = NULL;
-	uint64_t       cache_elem_size;
-
-	// Since we're not in an atomic context (see GFP_KERNEL flag above) this
-	// is an acceptable alternative to kmalloc however, it should be noted
-	// that this is a little less efficient. The reason for this is
-	// fragmentation that can occur on systems. We noticed this happening in
-	// the field, and if highly fragmented, our driver will fail to load
-	// with a normal kmalloc
+	int		    i;
+	uint64_t	    cache_elem_size;
+	size_t		    tableSize;
+	struct HashTbl *    hashTblp	  = NULL;
+	struct hashtbl_bkt *tbl_storage_p = NULL;
+
+	if (!is_power_of_2(numberOfBuckets)) {
+		numberOfBuckets = roundup_pow_of_two(numberOfBuckets);
+	}
+	printk(KERN_INFO "%s: %s bkts:%llu bits:%u", __func__, hashtble_name,
+	       numberOfBuckets, ilog2(numberOfBuckets));
+
+	hashTblp = kzalloc(sizeof(*hashTblp), GFP_KERNEL);
+	if (!hashTblp) {
+		return NULL;
+	}
+
+	tableSize = (numberOfBuckets * sizeof(struct hashtbl_bkt));
+
 	tbl_storage_p = vmalloc(tableSize);
 
-	if (tbl_storage_p == NULL) {
-		HASHTBL_PRINT("Failed to allocate %lluB at %s:%d.",
-			      (numberOfBuckets * sizeof(struct hlist_head)) +
-				      sizeof(struct HashTbl),
+	if (!tbl_storage_p) {
+		HASHTBL_PRINT("Failed to allocate %luB at %s:%d.", tableSize,
```

Comment:
Need to free the hashTblp before return.

Reply:
Good Catch! Thanks!